src/app/view/OsReceive/OsReceiveProvider.tsx

```diff
@@ -37,25 +38,31 @@ export const OsReceiveProvider = ({
   const [data, setQuery] = useState<AcceptFromFlagshipManifest[]>([])
   const didCall = useRef(false)
 
-  // This effect is triggered at mount and unmount of the provider,
-  // its role is to listen native events and update the state accordingly
   useEffect(() => {
-    // Pass a callback to the low level function that handles the received files
-    // We will have access to their paths in the provider state afterwards
-    const cleanupReceivedFiles = handleReceivedFiles(files => {
+    const onFilesReceived = (files: OsReceiveFile[]): void => {
+      if (!client?.isLogged) return
       dispatch({ type: OsReceiveActionType.SetFilesToUpload, payload: files })
-      backToHome().catch(err => {
-        OsReceiveLogger.warn(
-          'Error while going back to home after receiving files from native',
-          err
-        )
+      backToHome().catch(error => {
+        OsReceiveLogger.error('Could not go back to home', error)
       })
-    })
+    }
+
+    const onError = (error: unknown): void => {
+      OsReceiveLogger.error('Could not get received files', error)
+      dispatch({ type: OsReceiveActionType.SetFlowErrored, payload: true })
+    }
+
+    OsReceiveEmitter.on('filesReceived', onFilesReceived)
+    OsReceiveEmitter.on('error', onError)
+
+    OsReceiveEmitter.receiveFiles()
```

Comment:
Are you sure we should call `receiveFiles` from this useEffect? Seems like this will cancel all your refactoring purpose by calling `receiveFiles` multiples times.

Reply:
there is a race condition though on app start (event emitted before the provider can listen to it), I'm putting a `OsReceiveEmitter.ensureActivation()`

Reply:
And what about the `OsReceiveEmitter.clearReceivedFiles()` ?